add arrow keys on linux. #2

josdorrod opened the pull request:
Possibility of using arrows keys or asdw.
Modified number of movements.
Alphabetical arrangement of headers.

josdorrod committed:
add arrow keys on linux. #2

Possibility of using arrows keys or asdw.
Modified number of movements.
Alphabetical arrangement of headers.

user23052036 commented:
Thank you for contributing!
I will definitely look into it. But as you have made it system specific. Could try making it universal. Have any idea ?

josdorrod commented:
I could try to test on Windows.
I am not sure but I think that on macOS and Linux the behaviour is the same.

Test It and let me know your opinion.
Then i Will try to do the same for windows

user23052036 commented:
Okei sure.. I would look into it by tomorrow

user23052036 commented:
I creaked ur code works in linux fine.. But in windows its not working.. I have been also stuck in this problem trying to figure out a universal code independent of os.

josdorrod commented:
Is it valid if I look for an alternative to make it work for Windows and another for Linux? 
Then using variable in C we will know which OS is being used and take one path or another

user23052036 commented:
Yes sure that would work ...

user23052036 commented:
Take a look at the below code..
We could use #if #ellif preprocessor execution statement to find out about the os 

    #if defined(_WIN32) || defined(_WIN64)
        printf("Operating System: Windows\n");

    #elif defined(__APPLE__) && defined(__MACH__)
        printf("Operating System: macOS\n");

    #elif defined(__linux__)
        printf("Operating System: Linux\n");

    #elif defined(__unix__)
        printf("Operating System: Unix\n");

    #else
        printf("Operating System: Unknown\n");
    #endif

Kind of like this you can use it in algorithm.h function to determine the os and accordingly we can map the arrow keys to ASWD keys in algorithm.c

josdorrod committed:
Merge remote-tracking branch 'upstream/main' into arrow_keys

josdorrod committed:
Add get key pressed for windows.

It is no longer necessary to press enter after each keystroke in Windows.
Add new file getkey.c and header getkey.h and modified Cmake file.

josdorrod committed:
Add get_key for unix system

The code in main was alocated in getkey.c

user23052036 commented:
Okk.. give me 1 day time I will take a look at it by tomorrow

josdorrod commented:
I add key recognisition for windows (without having to push enter) and for unix system (pushing enter)
add .gitignore
Add getkey.c and getkey.h
Modified Cmake
Tested on windows and linux.
Let me know your opinion

user23052036 commented on code/getkey.h:
```diff
@@ -0,0 +1,7 @@
+#ifndef GETKEY_H
+#define GETKEY_H
+
+// Function to get the key pressed.
+char get_key_pressed(void);
+
+#endif
```
Have not included get_key() and get_key_window() function into the header file.

try this:-

`#ifndef GETKEY_H
#define GETKEY_H


char get_key_pressed(void);


// Function to get the key pressed system specific.
#if defined(_WIN32) || defined(_WIN64)

    #include<windows.h>
    #include <unistd.h>   //sleep func
    char get_key_window(void);

#else

    #include<ctype.h>
    #include <stdio.h>    //scanf
    char get_key(void);

#endif

#endif`

josdorrod replied:
I've read that functions that aren't used outside of the.c file where they're defined shouldn't go in a.h file. That's why i didn't include these functions.

user23052036 replied:
Well, for cross platform program its actually a good practice to do it.. To define all the header file and all the functions declarations in the header file.

user23052036 commented on code/getkey.c:
```diff
@@ -0,0 +1,70 @@
+#include <ctype.h>
```
good practice to always include all the header files and user defined header in .h files and write just code in .c file.            

Remove all the header files and just include getkey.h

josdorrod replied:
As I see it, in a .h file you should only add .h files that are directly related to the header file you are defining. 
In the example below the stddef.h library is needed because it is used for type size_t. 
The rest of the libraries should be declared in the .c file. 
The same thing happens with functions, just write the prototypes of the functions that are going to be used outside the .c file. 
At least that's what I've read as good practices declaring .c and .h files

```c
#ifndef MY_LIBRARY_H
#define MY_LIBRARY_H

#include <stddef.h>  // Need to use size_t

#define MAX_SIZE 100

size_t get_array_size(int arr[], size_t length);

#endif // MY_LIBRARY_H

user23052036 replied:
Yes you are right This is true when you are writing system specific code. 

But here as we are developing cross platform program to work.. including all the header in .c file will unnecessarily including extra header file like if i am running the game in windows i don't need (#include<ctype.h>  #include <stdio.h>) As similarly if i am in Linux system in would not need #include<windows.h> .. 

So, i have make the declarartion including functions declarations into header file.. So that it only includes what it needs

user23052036 commented on code/getkey.c:
```diff
@@ -0,0 +1,70 @@
+
+char get_key(void);
+
+char get_key_pressed(void)
+{
+    #if defined(_WIN32) || defined(_WIN64)
+        return get_key_window();
+    #else
+        return get_key();
+    #endif
+}
```
Control reaches end of non-void function. Please correct the code.

user23052036 commented on code/getkey.c:
Please make the code structure tha way i am sharing it below it more structured:-

`
#include"getkey.h"

#if defined(_WIN32) || defined(_WIN64)

    char get_key_window(void)
    {
        while (1) 
        {

            // 0x8000 compare the 15 bits. If it's 1, the key is pressed.
            if (GetAsyncKeyState(VK_UP) & 0x8000 || GetAsyncKeyState('W') & 0x8000) 
            {
                Sleep(50);
                return 'w'; // Evita múltiples detecciones
            }
            if (GetAsyncKeyState(VK_DOWN) & 0x8000 || GetAsyncKeyState('S') & 0x8000) 
            {
                Sleep(50);
                return 's';
            }
            if (GetAsyncKeyState(VK_LEFT) & 0x8000 || GetAsyncKeyState('A') & 0x8000) 
            {
                Sleep(50);
                return 'a';
            }
            if (GetAsyncKeyState(VK_RIGHT) & 0x8000 || GetAsyncKeyState('D') & 0x8000) 
            {
                Sleep(50);
                return 'd';
            }

            Sleep(50); // Para no sobrecargar la CPU
        }
    }


#else

    #include <ctype.h>    //toupper
    char get_key(void)
    {
        char key_pressed[5];
        scanf("%s",key_pressed);

        // Get key pressed.
        if (toupper(key_pressed[0]) == 'A' || toupper(key_pressed[0]) == 'S' || 
        toupper(key_pressed[0]) == 'D' || toupper(key_pressed[0]) == 'W')
            return key_pressed[0];

        // Arrow keys are converted into ^[[A, [[B, [[C, [[C on linux. ^[ = Esc.
        else if (toupper(key_pressed[0]) == 27 || toupper(key_pressed[1]) == '[')     // 27 = Esc
        {
            if (toupper(key_pressed[2]) == 'A')
                return 'w';
            else if (toupper(key_pressed[2]) == 'B')
                return 's';
            else if (toupper(key_pressed[2]) == 'C')
                return 'd';
            else if (toupper(key_pressed[2]) == 'D')
                return 'a';
        }
    }

#endif


char get_key_pressed(void)
{
    #if defined(_WIN32) || defined(_WIN64)
        return get_key_window();
    #else
        return get_key();
    #endif
}
`

user23052036 reviewed (commented):
Please correct the code that i have marked .. I cheaked in windows and sometimes when pressing the arrow keys a liitle long its shifting more than 1 number together.

josdorrod commented:
I'll review all the points you've made

user23052036 commented:
The logical part that you have written is correct. But when we are pressing the arrow key for a little long its shifting all the number in that directing and ending the game.

What we need to do is wait until the user stops pressing the arrow key and then return the character.

ex.. if we press arrow (left) for 5sec the program waits till i release the arrow key press.. and then it just retunes character 'a'...

josdorrod committed:
Add minor changes

Delete comment main.c
Reorganice getkey.c

josdorrod committed:
Modified get key pressed for window.

Add pointer direction as parameter to void function.
Wait until user stop pressing.

user23052036 commented:
Have you committed all the changes yet ?...Because I tried running and its not working.
Clicking on arrow key its not moving the Numbers
![Screenshot 2024-07-07 222515](https://github.com/user23052036/Puzzle-Game/assets/155566244/0706fdbb-d107-4008-b084-a0313c4de105)

josdorrod committed:
Improved the way to get the key pressed

Add validation only taking the first 4 characters in get_key.
Add waiting until key releasedin get_key_window.
Replace system(cls) for new function.

josdorrod commented:
Add minor change commented.
Add waiting key release on get_key_window.
Add validation to only take the first 4 characters on get_key.
Change system(cls) for new function that use cls or clear depending on the system.

user23052036 commented:
Great work.. Works good in windows i cheeked. Have you cheeked in Linux os ? Unfortunately for some reason my virtual engine turned off i am not able to use (Linux) right now.

josdorrod commented:
So It is working correctly?Yeah, I checked on Linux but I have to press enter after each key. There is a way to do without pressing enter but you have to change the Shell properties and i didn't want to risk.El 7 jul 2024 21:36, nomad2036 ***@***.***> escribió:
Great work.. Works good in windows i cheeked. Have you cheeked in Linux os ? Unfortunately for some reason my virtual engine turned off i am not able to use (Linux) right now.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

user23052036 commented:
Ok no prob.. I am merging it to the main branch

user23052036 commented:
I will try implementing the arrow key functionality for Linux
